### DISCUSSION TITLE
Aqualinkd PDA and EW-11 - RS485 issues

### DISCUSSION THREAD

**Discussion by sts098** [aqualinkd/AqualinkD#333] [General]
I am experiencing some connection issues.  I am hoping you may be able to help with a solution.  

BACKGROUND

Identifying Marks on the Main Board:
•	PCB#: E0260500D
•	ASSY #: E0260600W
•	White Sticker: B18RSCMG3821

And on the daughter-board:
•	B0029205E
•	B0029300C
•	White Sticker:  B0029205 / PDA / PS8 COMBO Rev 7.1

Logging in to iAqualink.net:
- interface: Aqualink Touch
- Model: PDA-PS8 Combo
- PDA Rev: PDA: 7.1.0
- CPU p/n B0029205
- TL Rev:

 
The main board has 2 RS485 ports.  The port on the right had three connections:
•	18 channel RF Transceiver – PDA
•	iAqualink 2 – ethernet (hard wired)
•	Jandy Variable Speed Pump

Left port was empty, so I initially connected my ELFIN EW11 to that port.  
No signal came through.  After a lot of playing with settings in the EW11, I finally connected it to the RS485 on right (fourth connection to this) and it worked.  

Aqualinkd does connect and I can see pool status and control the pump.  I have some issues connecting from time to time.  And noticed errors on aqmanager.html.

Serial Logger Results (from within aqmanager.html):
•	Notice:NetService:Received request to run serial_logger!
•	Warning:AqualinkD: Starting serial_logger, this will take some time!
•	Notice:RS Serial: Read 100 of 600 packets
•	Notice:RS Serial: Read 200 of 600 packets
•	Notice:RS Serial: Read 300 of 600 packets
•	Notice:RS Serial: Read 400 of 600 packets
•	Notice:RS Serial: Read 500 of 600 packets
•	Notice:RS Serial: Read 600 of 600 packets
•	Notice:RS Serial: RS485 interface received 600 packets in 12 seconds (~50.00 Msg/Sec)
•	Notice:RS Serial: Jandy Control Panel Model   :
•	Notice:RS Serial: Jandy Control Panel Version :
•	Notice:RS Serial: Jandy ID's found
•	Notice:RS Serial: ID 0x33 is in use    <-- Aqualink (iAqualink / Touch)
•	Notice:RS Serial: ID 0xa3 is in use    <-- Unknown Device
•	Notice:RS Serial: ID 0x60 is not used  <-- can use for Aqualinkd (PDA mode only)
•	Notice:RS Serial: ID 0x78 is in use    <-- Jandy VSP ePump
•	Notice:RS Serial:
•	Notice:RS Serial: Suggested aqualinkd.conf values
•	Notice:RS Serial: -------------------------
•	Notice:RS Serial: device_id = 0x60
•	Notice:RS Serial: -------------------------

The only changes I made to the default aqualinkd.conf include:
•	serial_port=/dev/tty.Pool2
•	panel_type = PD-8 Combo
•	device_id=0x60
•	button_01_pumpID=0x78
•	button_01_pumpIndex=1

I am getting a number of serial read warnings, both bad packets and bad checksums.  Additionally, I am getting the following errors: 
Error: PDA:       PDA Init :- wait for next menu failed

Error: PDA:       Could not get heater setpoints, trying again!
Error: PDA:       PDA Device programmer didn't find a requested menu 7, current 4
Error: PDA:       PDA Device programmer couldn't select menu item 'SET TEMP' menu 4

Error: PDA:       PDA Init :- Error getting freeze setpoints
Error: PDA:       PDA Device programmer didn't find a requested menu 15, current 12
Error: PDA:       PDA Device programmer couldn't select menu item 'FREEZE PROTECT' menu 12

Error: PDA:       waitForPDAMessageTypes: did not receive 0x08 or 0x04
Error: PDA:       waitForPDAMessageTypes: did not receive 0x08 or 0x04

Error: PDA:       PDA Init :- can't find menu
Error: PDA:       loopover_devices :- can't find ALL OFF

Do you think the errors are from bad packets or do I have some incorrect settings?

I have read that each of the Jandy RS485 ports should have only 2 connections.  Mine FROM INSTALLER had 3.  With the EW11, I am up to 4.  Do you think the Jandy 6584 Multiplexer may solve the BAD PACKET problem?  

Do you have any idea why the left most RS485 isn’t functioning?  Are there any dip switch settings, etc.  Or is it likely because of the PDA CPU.

Do you have any suggestions for me?  

I do not know if I should jump on a similar conversation as someone is actively discussing EW11 issues with you.  So I started this thread.  

Thanks for all the hard work you have done to make this.

**Comment by sfeakes** [comment]
No idea why the 2nd RS485 port is not working.  They are literally hardwired together on the board, so putting 2 sets of wires in each terminal block is identical to putting 4 sets of wires on one terminal.  The Jandy multiplexer board is identical in that it's simply hardwiring more terminals nothing special to it. (It's a waste of money and won't help you, unless you physically can't stuff the wires in the terminal block.)
The red RS485 terminal block is removable, so maybe make sure it's seated correctly.

PDA boards are slower and way more error prone that RS boards, AqualinkD has to sent a series of 10 or so commands to do the simplest task on a PDA board, and if their is any missed messages / checksum error in that series, it's not going to work. On an RS board, most things can be done by sending 1 command.  So until you get the bad backed stuff sorted, you won't get far on a PDA board.
In your aqualinkd.conf one thing that may help is the 'rs485_frame_delay = 4' setting.  4 is 4 milliseconds, PDA boards bing slower that RS boards sometimes benefit from changing that to make AqualinkD reply slower. Usually I'd say increase that to 8 or so, but the EW11 is almost certainly adding some latency, so maybe set that to 0.

Other than that, look at the EW11 settings and WiFi settings to make sure that it not adding any issues.

**Comment by sts098** [comment]
Thanks for getting back to me.  

SERIAL PORT: 
Do you know if there is any way to test the RS485 port?  Should a multimeter read a voltage across Terminal 4 and terminal 1?  How about across the data pins 2 and 3?  

PDA vs RS Board
As you may guess, I am trying to avoid the costs of replacing the CPU board.  I am concerned with cost/system downtime/programming.  I may consider at some time, but more as a last resort. 

FRAME DELAY: 
I am playing with the rs485_frame_delay settings.  Seems to improve at 2, trying others.  It doesn't get rid of them, maybe improves them.  I will keep investigating.  Is there any way to quantify results?  Can I Record serial errors and the time between them to determine improvements?

EW11 SETTINGS:
On EW11 serial port, I am running 9600/8/n/1 half duplex, which I believe all to be ok.  There are 2 settings buffer and gap.  
•	Currently running a buffer size of 512, seen both 512 and 1024 listed by others.  Do You have a recommendation?  
•	On The Gap setting, Help lists this as UART free-frame gap time in ms, I have the default 50.  Do you have any thoughts here? Could this be affecting timing/Frame Delay?  

On EW11 Socket Settings, I am running a buffer of 512 (same as the UART Buffer), keep alive of 60 s, and a time out of 300.  

My connection times can get excessive on restart.  Could this be due to consuming buffered information?  Could large buffers be causing syncing issues on start up?  Or am I misunderstanding their function.  Anyway, do you have any suggested buffer sizes for the UART and Socket?  

Thanks again.

UPDATE:
Changing UART free-frame gap time to 10ms (minimum value for the EW11) and setting rs485_frame_delay to 0 took me from about 5-6 errors per minute to around (8-12 over 10 minutes).

**Reply by sfeakes** [reply]
It sounds like GAP and Frame Delay might be the same thing. Or compounding the issue.  On Jandy, the panel will send a message and expect a reply within 40ms. If it doesn’t get that reply, then it considers that device dead. AqualinkD will usually send a reply within 1ms (But that will depend on HW used). 1ms can be too quick for PDA panels, and cause problems. That’s why the rs485_frame_delay was put in. The default value or 4 is 4ms. 
I’m not sure how the EW11 implements GAP, it could be simply space between receiving packets, ie if it doesn’t retrieve a packet within 50ms then consider the  frame complete and send the frame.  Or it could be wait 50ms before sending anything.  Or maybe a combination of both.  But considering you reduced the errors by reducing GAP, I’d think it may be the latter.  I’d leave AqualinkD frame delay at 0, and bring EW11 gap to 4ms and see what happens.

There are a couple of people playing with EW11 at the moment, so not much is known about best settings for Jandy.  But you are the first with a PDA panel, and as stated that has it’s own set of challenges.

Edit
512 buffer should be enough, AqualinkD uses 128.
9600/8/n/1 half duplex - also looks good.

**Reply by sts098** [reply]
I cannot reduce the gap setting below 10 (minimum).  Since reducing Gap setting, Aqualinkd connects immediately, no more long delays.  I still get bad checksums every minute or so, but only know this because I am running aqmanager.html.  I can successfully read the status and control Pumps, cleaner, lights, etc.  I will play with wiring to see if it is a bad connection.  

TEMP SETPOINTS:
I am getting some bad settings:
Panl&Prog: requested setpoint value 219 is invalid, change to 104
Panl&Prog: Setpoint of 219 for Spa is outside range, using 104
Panl&Prog: about to overwrite unactioned panel program
Panl&Prog: requested setpoint value 158 is invalid, change to 104
Panl&Prog: Setpoint of 158 for Pool is outside range, using 104

219 and 158 are not the actual values in my system.  They are 104 and 80.  Not sure why it is being reported incorrectly.  

These errors are consistent, every time I boot up or restart aqualinkd, so i do not believe it is a bad bit.  So I ma curious if PDA Rev 7.1.0 is storing the vale differently.  

AQUALINK TOUCH PROTOCOL:
I am also curious, Aqualinkd logger can see my IAqualink Touch interface.  
- ID 0x33 is in use    <-- Aqualink (iAqualink / Touch)   
It is also shown on Jandy webportal https://iaqualink.zodiacpoolsystems.com/
- Interface: AquaLink Touch

But when I try to enable, I get the message:
AqualinkD: Aqualink daemon can't use extended_device_id in PDA mode, ignoring value '0x31' from cfg

Is this a hardware issue, or is it turned off in the software?  Anyway to enable it?

**Comment by sfeakes** [comment]
AqualinkD can not use 0x31 in PDA mode. I’m slowly working on that, but it’s not finished. The serial logger has been updated in the current dev release to make that message less confusing.

As for the 219 and 158 something is requesting AqualinkD to set those valves.  Do you have an integration (maybe MQTT) that’s making that request?  If you turn on debugmask for net services it will tell you if that request is from MQTT or Webservice.

**Comment by sfeakes** [comment]
Moving EW-11 discussions to one place.
https://github.com/sfeakes/AqualinkD/discussions/336
